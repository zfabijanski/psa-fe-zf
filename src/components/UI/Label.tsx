import React from "react";
import styled, { CSSProp } from "styled-components/macro";
import { useIntl } from "react-intl";
import { ModalInfoIcon } from "./ModalInfoIcon";
import { IModalData } from "../../models/Modal";
import { TypoSmallRegular } from "./typography";
import { TranslationValue } from "utils/types";

interface IInnerIconsCoordinates {
  right: string;
  top: string;
}

export interface ILabelProps extends IModalData {
  labelTextStyles?: CSSProp;
  isInvalid?: boolean;
  labelText?: string | React.ReactNode;
  labelTrKey?: string;
  innerIcons?: boolean;
  validationInfoTrKeys?: {
    trKey: string;
    values?: Record<string, TranslationValue>;
  }[];
  handleLabelClick?: () => void;
  innerIconsCoordinates?: IInnerIconsCoordinates;
  className?: string;
  hasDarkBackground?: boolean;
}

export const Label: React.FC<ILabelProps> = (props) => {
  const {
    labelTextStyles,
    labelText,
    labelTrKey,
    innerIcons,
    modalContentTrKey,
    modalContent,
    modalButtons,
    handleLabelClick,
    innerIconsCoordinates,
    className,
    modalHtml,
    modalHtmlTrValues,
    hasDarkBackground,
  } = props || {};

  const intl = useIntl();

  return (
    <LabelContainer
      labelWithText={!!(labelTrKey || labelText)}
      className={className}
    >
      <LabelText labelTextStyles={labelTextStyles} onClick={handleLabelClick}>
        {labelText ||
          (labelTrKey ? intl.formatMessage({ id: labelTrKey }) : "")}
      </LabelText>
      <Icons
        innerIcons={innerIcons}
        innerIconsCoordinates={innerIconsCoordinates}
      >
        {(modalContentTrKey || modalHtml) && (
          <ModalInfoIcon
            modalContentTrKey={modalContentTrKey}
            modalContent={modalContent}
            modalButtons={modalButtons}
            modalHtml={modalHtml}
            modalHtmlTrValues={modalHtmlTrValues}
            hasDarkBackground={hasDarkBackground}
          />
        )}
      </Icons>
    </LabelContainer>
  );
};

Label.defaultProps = {
  innerIconsCoordinates: {
    top: "calc(100% + 22px)",
    right: "calc(100% - 26px)",
  },
};

const LabelContainer = styled.div<{ labelWithText: boolean }>`
  position: relative;
  ${({ labelWithText }) => labelWithText && "margin-bottom: 4px"};
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const LabelText = styled(TypoSmallRegular)<{ labelTextStyles?: CSSProp }>`
  cursor: default;
  display: inline-block;
  width: 100%;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: pre;
  ${({ labelTextStyles }) => labelTextStyles}
`;

const Icons = styled.div<{
  innerIcons?: boolean;
  innerIconsCoordinates?: IInnerIconsCoordinates;
}>`
  display: flex;
  margin-bottom: 5px;
  z-index: 1;

  ${({ innerIcons, innerIconsCoordinates }) =>
    innerIcons &&
    `
    position: absolute;
    top: ${innerIconsCoordinates?.top};
    right: ${innerIconsCoordinates?.right};
  `}

  > *:not(:last-child) {
    margin-right: 10px;
  }
`;
