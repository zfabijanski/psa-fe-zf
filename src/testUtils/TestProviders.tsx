import { render, RenderOptions } from "@testing-library/react";
import { FC, PropsWithChildren, ReactElement } from "react";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import { Normalize } from "styled-normalize";
import GlobalStyles from "theme/globalStyles";
import { theme } from "theme/theme";
import rootStore, { configureStore, rootReducer as reducer } from "AppStore";
import { BrowserRouter } from "react-router-dom";
import { Modal } from "components/UI/Modal";
import { IntlProvider } from "react-intl";
import { plMessages } from "translations/pl";
import I18nProvider from "I18nProvider";

interface TestProvidersProps {
  store?: typeof rootStore;
  renderModal?: boolean;
  realIntlProvider?: boolean;
}

export const addTestProviders =
  ({
    store = configureStore({ reducer }),
    renderModal = true,
    // @TODO: update tests to use realIntlProvider everywhere
    realIntlProvider = false,
  }: TestProvidersProps): FC<PropsWithChildren<{}>> =>
  ({ children }) => {
    const TranslationsProvider = realIntlProvider ? IntlProvider : I18nProvider;

    return (
      <Provider store={store}>
        <BrowserRouter>
          <ThemeProvider theme={theme}>
            <TranslationsProvider locale="pl" messages={plMessages}>
              <Normalize />
              <GlobalStyles />
              {children}
              {renderModal && <Modal />}
            </TranslationsProvider>
          </ThemeProvider>
        </BrowserRouter>
      </Provider>
    );
  };

const customRender = (
  ui: ReactElement,
  options: Omit<RenderOptions, "options"> = {},
  testProviderOptions: TestProvidersProps = {}
) => render(ui, { wrapper: addTestProviders(testProviderOptions), ...options });

export * from "@testing-library/react";

// override render method
export { customRender as render };
