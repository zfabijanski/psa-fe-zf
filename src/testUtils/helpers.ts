export const sleep = (ms: number = 100): Promise<unknown> =>
  new Promise((resolve) => setTimeout(resolve, ms));
