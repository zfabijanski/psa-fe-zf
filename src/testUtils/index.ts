export * from "./TestProviders";
export * from "./mockCreators";
export * from "./helpers";
