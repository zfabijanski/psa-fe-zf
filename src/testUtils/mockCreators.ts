import {
  ICalculatorCoversConfig,
  IProductCovers,
} from "features/Calculator/types";
import {
  DocumentState,
  IDirectory,
  IDirectoryMetadata,
  IDocumentMetadata,
} from "features/Library/types";
import { IMeeting } from "features/Meeting/History/types";
import { IIllustration, ProductGuid } from "features/Products/types";
import { IModalData, ModalTypes } from "models/Modal";
import {
  Frequency,
  IFundRiskProfileConfig,
  IProductConfig,
  IProductFund,
  IProductGuarantee,
  InvestmentProfile,
  Role,
} from "models/common";
import {
  CalculationStatus,
  Direction,
  ICalculationDetailsDto,
  ICover,
  IInsured,
  YesNo,
} from "models/calculator";

export const createCalculationDetailsDto = (
  overrides?: Partial<ICalculationDetailsDto>
): ICalculationDetailsDto => ({
  calculation_id: 1,
  working_date: "2020-01-01",
  premiums_total: 100,
  guaranteed_death_benefit: 100,
  guaranteed_maturity_benefit: 100,
  projected_maturity_value_egr_m: 100,
  product_categories: [],
  uuid: "uuid",
  version: "version",
  source_system: "source_system",
  product_guid: 1,
  insureds_list: [],
  covers_list: [],
  funds_list: [],
  ...overrides,
});

export const createInsured = (overrides?: Partial<IInsured>): IInsured => ({
  insured_idx: 1,
  insured_uuid: "insured_uuid",
  age_in_months: 1,
  date_of_birth: "2020-01-01",
  role: Role.MAIN,
  ...overrides,
});

export const createCover = (overrides?: Partial<ICover>): ICover => ({
  cover_id: 1,
  cover_idx: 1,
  insured_idx: 1,
  code: "code",
  frequency_code: Frequency.Annually,
  payments: 1,
  direction_of_calculation: Direction.SumAssuredToPremium,
  investment_profile: InvestmentProfile.Balanced,
  indexation: YesNo.Yes,
  sum_assured: 1,
  premium_per_freq: 1,
  duration: 1,
  idx_option: YesNo.Yes,
  rop_option: YesNo.Yes,
  add_sum_assured: 1,
  ending_age: 1,
  exp_mat_ben: 1,
  ...overrides,
});

export const createCalculatorCoversConfig = (
  overrides?: Partial<ICalculatorCoversConfig>
): ICalculatorCoversConfig => ({
  coverName: "coverName",
  isChild: false,
  isHospital: false,
  ...overrides,
});

export const createCalculatorCoversConfigMap = (
  overrides?: Partial<Record<string, ICalculatorCoversConfig>>
): Record<string, ICalculatorCoversConfig> => ({
  coverName1: createCalculatorCoversConfig({ coverName: "coverName1" }),
  coverName2: createCalculatorCoversConfig({ coverName: "coverName2" }),
  ...overrides,
});

export const createProductCovers = (
  overrides?: Partial<IProductCovers>
): IProductCovers => ({
  mainCover: "mainCover",
  additionalCovers: ["additionalCover1", "additionalCover2"],
  wopCovers: ["wopCover1", "wopCover2"],
  childCovers: ["childCover1", "childCover2"],
  ...overrides,
});

export const createProductCoversMap = (
  overrides?: Partial<Record<string, IProductCovers>>
): Record<string, IProductCovers> => ({
  mainCover1: createProductCovers({ mainCover: "mainCover1" }),
  mainCover2: createProductCovers({ mainCover: "mainCover2" }),
  ...overrides,
});

export const createProductConfig = (
  overrides?: Partial<IProductConfig>
): IProductConfig => ({
  product_guid: 1,
  product_name: "product_name",
  product_categories: [],
  product_freqs: [],
  product_covers: [],
  product_funds: [],
  ...overrides,
});

export const createProductConfigMap = (
  overrides?: Partial<Record<string, IProductConfig>>
): Record<string, IProductConfig> => ({
  1: createProductConfig({ product_name: "product_name1" }),
  2: createProductConfig({ product_name: "product_name2" }),
  ...overrides,
});

export const createFundRiskProfileConfig = (
  overrides?: Partial<IFundRiskProfileConfig>
): IFundRiskProfileConfig => ({
  code: "code",
  name_pl: "name_pl",
  name_en: "name_en",
  risk_level: 1,
  status_code: "status_code",
  notes: "notes",
  ...overrides,
});

export const createProductGuarantee = (
  overrides?: Partial<IProductGuarantee>
): IProductGuarantee => ({
  guarantee_code: "guarantee_code",
  guarantee_name: "guarantee_name",
  guarantee_min_value: "1",
  guarantee_duration_months: "4",
  guarantee_percentages: [{ guarantee_percent: 10 }],
  ...overrides,
});

export const createProductFund = (
  overrides?: Partial<IProductFund>
): IProductFund => ({
  cover_code: "cover_code",
  fund_code: "fund_code",
  fund_name: "fund_name",
  fund_risk_profile: "code",
  fund_order: 1,
  guarantees: [createProductGuarantee()],
  ...overrides,
});

export const createModalData = (
  overrides: Partial<IModalData> = {}
): IModalData => ({
  modalIllustration: "alert-fill",
  modalTitleTrKey: "modalTitleTrKey",
  modalContentTrKey: "modalContentTrKey",
  modalContent: "modalContent",
  modalButtons: {
    confirm: { textTrKey: "modal.ok" },
  },
  modalType: ModalTypes.default,
  maskClosable: true,
  ...overrides,
});

export const createIllustration = (
  overrides: Partial<IIllustration> = {}
): IIllustration => ({
  idd_id: 1,
  product_guid: ProductGuid.Savings,
  adequacy_id: 1,
  order_id: 1,
  expected_amount: { value: 3 },
  premium_total: { value: 3 },
  duration: { value: 3 },
  premium_main_cover: { value: 3 },
  frequency: { value: 3 },
  sum_of_premiums: { value: 3 },
  sum_assured: { value: 3 },
  investment_profile: { value: 3 },
  pru_calc_calculation_id: 1,
  status: CalculationStatus.Illustrated,
  json_needs_analysis: null,
  ...overrides,
});

export const createDirectory = (
  overrides?: Partial<IDirectory>
): IDirectory => ({
  directoryId: 1,
  directories: [],
  documentsIds: [1, 2, 3],
  ...overrides,
});

export const createDirectoryMetadata = (
  overrides?: Partial<IDirectoryMetadata>
): IDirectoryMetadata => ({
  directoryId: 1,
  name: "name",
  isOpen: false,
  ...overrides,
});

export const createDocumentMetadata = (
  overrides?: Partial<IDocumentMetadata>
): IDocumentMetadata => ({
  documentId: 1,
  guid: "guid",
  name: "name",
  state: DocumentState.NotAttached,
  isStateChanging: false,
  moduleName: "PRU_LIBRARY",
  ...overrides,
});

export const createMeeting = (overrides?: Partial<IMeeting>): IMeeting => ({
  meetingNo: "meetingNo",
  meetingId: 1,
  name: "name",
  age: 18,
  lastAccessDate: new Date("2023-01-01"),
  createDate: new Date("2023-01-01"),
  statusId: 1,
  statusName: "statusName",
  statusDescription: "statusDescription",
  state: "state",
  haveAdequacy: false,
  haveCalculation: false,
  haveIdd: false,
});
