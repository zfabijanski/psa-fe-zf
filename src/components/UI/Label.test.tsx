import { Label, ILabelProps } from "./Label";
import { render, screen } from "testUtils";

const setup = (props: ILabelProps) => render(<Label {...props} />);

describe("Label", () => {
  it("should render the label", () => {
    setup({ labelText: "label" });
    expect(screen.getByText("label")).toBeInTheDocument();
  });

  it("should render the label with text styles", () => {
    setup({ labelText: "label", labelTextStyles: { color: "red" } });
    expect(screen.getByText("label")).toHaveStyle("color: red");
  });

  it("should render label with translation key", () => {
    setup({ labelTrKey: "labelTrKey" });
    expect(screen.getByText("labelTrKey")).toBeInTheDocument();
  });

  it("should render modal info icon", () => {
    setup({ modalContentTrKey: "modalContentTrKey" });
    expect(screen.getByText("info.svg")).toBeInTheDocument();
  });

  it("should render modal info icon with html", () => {
    setup({ modalHtml: "modalHtml" });
    expect(screen.getByText("info.svg")).toBeInTheDocument();
  });

  it("should render inner icons", () => {
    const { container } = setup({ innerIcons: true });
    expect(container).toMatchSnapshot();
  });
});
