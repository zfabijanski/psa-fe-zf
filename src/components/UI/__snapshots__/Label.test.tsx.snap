// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label should render inner icons 1`] = `
.c1 {
  font-weight: 400;
  font-size: 14px;
  line-height: 20px;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  cursor: default;
  display: inline-block;
  width: 100%;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: pre;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 5px;
  z-index: 1;
  position: absolute;
  top: calc(100% + 22px);
  right: calc(100% - 26px);
}

.c3 > *:not(:last-child) {
  margin-right: 10px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    />
    <div
      class="c3"
    />
  </div>
  <div />
</div>
`;
